README
MonitoringPlug
==============

My collection of monitoring plugins for Nagios and similar monitoring systems. 

Base
----

Basic plugins which don't use any library.

*  check_bonding -- Check bonding status.
*  check_dhcp -- Check DHCP server for functionality.
*  check_file -- Check the property of a file.
*  check_gsm_signal -- Check the signal quality of a GSM modem.
*  check_mem -- Check memory usage.
*  check_memcached -- Check memcached status.
*  check_multipath -- Check multipath for failed paths.
*  check_nrped -- Check if run inside of nrpe.
*  check_sockets -- Check socket count (Linux only). 

CUPS
----

Use CUPS to query CUPS server.

*  check_cups_jobs -- Check CUPS job count and age. 

CURL
----

Use the libcurl to test network servers.

*  check_apache_status -- Check apache mod_status output.
*  check_aspsms_credits -- Check available ​ASPSMS credits.
*  check_buildbot_slave -- Check BuildBot slave state by json-api.
*  check_tftp -- Check if a file can be downloaded from tftp.
*  check_webdav -- Check a WebDAV share. 

DNS
---

Use the ldns library to query dns servers.

*  check_dns_authoritative -- Check a Authoritative DNS server.
*  check_dns_sync -- Check if a DNS zone is in sync.
*  check_dnssec_expiration -- Check if a DNS zone signatur is not expired.
*  check_dnssec_trace -- Check if a DNS zone signatur is traceable from a trust anchor.
*  check_dnssec_trust_anchor -- Check if the trust anchors in named.conf are valid.

GNUTLS
------

Use the GnuTLS library to check SSL/TLS related things.

*  check_ssl_cert -- Check expiration and trust of a SSL certificate. 

IPMI
----

Use the OpenIPMI library to check BMC Sensors.

*  check_ipmi_fan -- Check the give or all FANs by IPMI.
*  check_ipmi_mem -- Check Memory status by IPMI.
*  check_ipmi_psu -- Check one or all PSU by IPMI sensor.
*  check_ipmi_sensor -- Check the give or all IPMI Sensors. 

LibVirt
-------

Use the LibVirt library to check VirtualMachines.

*  check_libvirtd -- Check the libvirtd itself.
*  check_libvirt_domain -- Check if a given domain is running. 

MySQL
-----

Use libmysqlclient library to check a MySQL Server.

*  check_mysql -- Check MySQL connectivity and status.
*  check_mysql_rows -- Check mysql table row count. 

RHCS
----

Plugins to check state of a RedHat Cluster Suite.

*  check_clustat -- Parse the clustat output. (Requires expat.)
*  check_rhcsnmp -- Check the state of a RedHat Cluster Suite by snmp. 

(SUN)RPC
--------

Plugins to check SUNRPC service state.

*  check_nfs -- Check if the Host is exporting at least one or the named path.
*  check_rpc_ping -- Check if named RPC program is responding. 

SELinux
-------

Check SELinux factors.

*  check_enforce -- Check SELinux state and policy.
*  check_sebool -- Check SELinux boolean state. 

SMB
---

Check SMB/CIFS.

*  check_smb_share -- Check SMB/CIFS share connection. 

SNMP
----

Use the ​net-snmp library to query SNMP agents.

*  check_akcp -- Check the state of a AKCP environment sensor.
*  check_apc_pdu -- Check the psu and outlet status of a APC PDU.
*  check_arc_raid -- Check the raid status of a Areca RAID.
*  check_interface -- Check Interface state with SNMP IF-MIB.
*  check_qnap_disks -- Check the dist status of a QNap.
*  check_qnap_vols -- Check the volume status of a QNap.
*  check_snmp_ups -- Check status of a UPS conforming to RFC 1628 by SNMP. 

XMLRPC
------

Use the xmlrpc-c library to query XMLRPC server.

*  check_koji_builder -- This plugin check a Koji-Builder.
*  check_koji_hub -- This plugin check a Koji-Hub.
*  check_rhn_entitlements -- Check available entitlement on a RedHat Satelite or RHN. 

Dummy
-----

*  check_dummy -- Don't check anything, return arguments.
*  check_timeout -- Don't check anything, cause a timeout. 

Notify
------

Some notifications plugins for Nagios.

*  notify_aspsms -- Send a notification by SMS with a ASPSMS account.
*  notify_mail -- Send a notification by mail.
*  notify_sms -- Send a notification by SMS with a Modem.
*  notify_stdout -- Print a notification to stdout for debuging. 

Enjoy!
  Marius


debian/changelog
monitoringplug (0.17.0) monitoringplug; urgency=low

  * Rework autotools build system.
  * Allow % sign in ranges and parse them.
  * Use hiredis library for redis checks.
  * Use long for redis memory size.
  * Add check_redis_slave check.
  * Add RabbitMQ checks.
  * Add LDAP checks.
  * Add Keepalived VRRP check.

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sat, 02 Aug 2014 11:45:05 +0200

monitoringplug (0.16.1) monitoringplug; urgency=low

  * Make .deb more interoperable

 -- Marius Rieder <marie@pcmarie.nine.ch>  Tue, 15 Oct 2013 17:34:17 +0200

monitoringplug (0.16) monitoringplug; urgency=low

  * Added PostgreSQL checks
  * Added FastCGI checks
  * Added Varnish check
  * Improve Redis check
  * Added SNI support to check_ssl_cert
  * Some bugfixes.


 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sun, 08 Sep 2013 09:43:46 +0200


monitoringplug (0.15) monitoringplug; urgency=low

  * Added check_redis

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sun, 02 Jun 2013 13:24:39 +0200

monitoringplug (0.14) monitoringplug; urgency=low

  * Added check_oping
  * Added check_dns_authoritative      
  * Hidden master support for check_dns_sync      
  * Added check_memcached 
  * Fix some minor bugs

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sat, 01 Dec 2012 18:04:22 +0200

monitoringplug (0.13) monitoringplug; urgency=low

  * Added check_snmp_usv check (Credits to Oliver Schonefeld)
  * Some snmp bugfixed
  * check_ipmi_*: Ignore unreadable sensors.
  * Build system improvments.

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Tue, 31 Jul 2012 11:55:23 +0200

monitoringplug (0.12) monitoringplug; urgency=low

  * Fix a SNMP, eopt, perfdata and clustat bug.
  * Improved check_akcp.
  * Added STARTTLS support to check_ssl_cert

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sun, 23 Jul 2012 17:55:23 +0200

monitoringplug (0.11) monitoringplug; urgency=low

  * Added IPMI checks with OpenIPMI.
  * Added GSM-Modem signal quality check.
  * Added SMS and mail notification command.

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sat, 14 Jul 2012 16:51:23 +0200

monitoringplug (0.10) monitoringplug; urgency=low

  * Initial DEB Release.

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Tue, 22 May 2012 20:06:46 +0200


lib/mp_args.h
/***
 * Monitoring Plugin - mp_args.h
 **
 *
 * Copyright (C) 2012 Marius Rieder <marius.rieder@durchmesser.ch>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * $Id$
 */

#ifndef _MP_ARGS_H_
#define _MP_ARGS_H_

#include "mp_eopt.h"

#include <stdint.h>

/**
 * prototype for argument parsing function.
 * Needs to be implemented by checks.
 * \param[in] argc number of arguments
 * \param[in] argv list of argumen strings 
 * \return \ref OK or \ref ERROR.
 */
int process_arguments(int argc, char **argv);

/**
 * Defining a range with start and end
 */
typedef struct range_struct {
    double  start;              /**< start of range.     */
    uint8_t start_infinity;     /**< start in -infinity  */
    uint8_t start_percent;      /**< start is in percent */
    double  end;                /**< end of range        */
    uint8_t end_infinity;       /**< end in infinity     */
    uint8_t end_percent;        /**< end is in percent   */
    uint8_t alert_on;           /**< invert range        */
} range;

/**
 * Bundle two ranges to a trashold
 */
typedef struct thresholds_struct {
    range   *warning;           /**< warning range      */
    range   *critical;          /**< critical range     */
} thresholds;

/**
 * Helper enum for \ref range_struct.
 * Used for bether readability of \ref range_struct.alert_on
 */
enum {
    OUTSIDE = 0,                /**<  Match if outside of range */
    INSIDE = 1                  /**<  Match if inside of range  */
};

/**
 * Helper enum for multiplier
 * Used for bether readability in \ref setWarn, \ref setCrit
 */
enum {
    NOEXT = -1,                 /**< No Extension                       */
    BISI = 0,                   /**< Binary/SI Prefix. Use
                                     \ref parse_multiplier_string       */
    TIME = 1                    /**< Times. Use
                                     \ref parse_time_multiplier_string  */
};

/**
 * Helper enum for long only opts
 */
enum {
    ARG_PERFDATA = -1
};

/**
 * Set the warning range of the given trashold.
 * If the thresholds or the warning range of the thresholds is NULL,
 * the needed memory is allocated.
 * \param[out] threshold Trashold pointer to write in.
 * \param[in] str Range string to parse.
 * \param[in] multiplier Multiplier syntar/function to use.
 * \return \ref OK or \ref ERROR.
 */
int mp_threshold_set_warning(thresholds **threshold, const char *str, int multiplier);

/**
 * Set the critical range of the given trashold.
 * If the thresholds or the critical range of the thresholds is NULL,
 * the needed memory is allocated.
 * \param[out] threshold Trashold pointer to write in.
 * \param[in] str Range string to parse.
 * \param[in] multiplier Multiplier syntar/function to use.
 * \return \ref OK or \ref ERROR.
 */
int mp_threshold_set_critical(thresholds **threshold, const char *str, int multiplier);

/**
 * Set the warning range of the given trashold using the time multiplier
 * function.
 * If the thresholds or the warning range of the thresholds is NULL,
 * the needed memory is allocated.
 * \param[out] threshold Trashold pointer to write in.
 * \param[in] str Range string to parse.
 * \return \ref OK or \ref ERROR.
 */
int mp_threshold_set_warning_time(thresholds **threshold, const char *str);

/**
 * Set the critical range of the given trashold using the time multiplier
 * function.
 * If the thresholds or the critical range of the thresholds is NULL,
 * the needed memory is allocated.
 * \param[out] threshold Trashold pointer to write in.
 * \param[in] str Range string to parse.
 * \return \ref OK or \ref ERROR.
 */
int mp_threshold_set_critical_time(thresholds **threshold, const char *str);

void free_threshold(thresholds *threshold);

/**
 * Parse a string into a range.
 *
 * Range String Syntax: [@][START:]END
 *
 * - "@" at the beginning makes the range inside out.
 * - START is 0 if not definied.
 * - START can be "~" for inf.
 * - END can be "~" or "" for inf.
 *
 * \param[out] range The range to parse into.
 * \param[in] str String to parse
 * \param[in] multiplier Multiplier syntar/function to use.
 */
int parse_range_string(range *range, const char *str, int multiplier);

/**
 * Parse a string into a multipliert for SI and Binary extensions.
 * - k  => 1'000                      K  => 1'024
 * - m  => 1'000'000                  M  => 1'048'578
 * - g  => 1'000'000'000              G  => 1'073'741'824
 * - t  => 1'000'000'000'000          T  => 1'099'511'627'776
 * - p  => 1'000'000'000'000'000      P  => 1'125'899'906'842'624
 * - e  => 1'000'000'000'000'000'000  E  => 1'152'921'504'606'846'976
 * \param[in] str String to parse.
 * \return Multipliert for the string or 1.
 */
double parse_multiplier_string(char *str);

/**
 * Parse a string into a multipliert for time extensions.
 * - s,sec => 1
 * - m,min => 60
 * - h,hr  => 3600
 * - d,day => 86400
 * - w,week => 604800
 * \param[in] str String to parse.
 * \return Multipliert for the string or 1.
 */
double parse_time_multiplier_string(char *str);

/**
 * Check of a value is inside a range.
 * \param[in] value Value to check.
 * \param[in] my_range Range to check.
 * \return 1 if value is in my_range, 0 otherwise.
 */
int check_range(double value, range *my_range);

/**
 * Get the state for a value, trashold combination.
 * \param[in] value Value to check.
 * \param[in] my_thresholds Trashold to check.
 * \return \ref STATE_CRITICAL if value is in critcal range of trashold,
 * \ref STATE_WARNING if value is in warning range of trashold only and
 * \ref STATE_OK otherwise.
 */
int get_status(double value, thresholds *my_thresholds);

/**
 * Prints a human-readable version of the my_threshold to stdout for
 * debuging  purpose.
 * \param[in] threshold_name Name of the trashold.
 * \param[in] my_threshold Trashold to print.
 */
void print_thresholds(const char *threshold_name, thresholds *my_threshold);

/**
 * Return a range string,
 * \param[in] my_range Range return string for.
 * \return string representing range.
 */
char *str_range(range *my_range);

/**
 * Prints the help message for the defaults options.
 */
void print_help_default(void);

/**
 * Prints the notification help message for the defaults options.
 */
void print_help_notify(void);

/**
 * Prints the help for the host option
 */
void print_help_host(void);

/**
 * Prints the help for the port option.
 * \param[in] def The default string.
 */
void print_help_port(const char *def);

/**
 * Print the help for a warning time option.
 * \param[in] def The default string.
 */
void print_help_warn_time(const char *def);

/**
 * Print the help for a critical time option.
 * \param[in] def The default string.
 */
void print_help_crit_time(const char *def);

/**
 * Print the help for a warning option.
 * \param[in] def The default string.
 */
void print_help_warn(const char *limit, const char *def);

/**
 * Print the help for a critical option.
 * \param[in] def The default string.
 */
void print_help_crit(const char *limit, const char *def);

/**
 * Print the help for the IPv4/6 option.
 */
void print_help_46(void);

/**
 * Parse the option for host.
 * \param[in] optarg option argument
 * \param[out] hostname hostname variable to set
 */
void getopt_host(const char *optarg, const char **hostname);

/**
 * Parse the option for host. Allow only IPs.
 * \param[in] optarg option argument
 * \param[out] hostname hostname variable to set
 */
void getopt_host_ip(const char *optarg, const char **hostname);

/**
 * Parse the option for port.
 * \param[in] optarg option argument
 * \param[out] port port variable to set
 */
void getopt_port(const char *optarg, int *port);

/**
 * Parse the option for url.
 * \param[in] optarg option argument
 * \param[out] url url variable to set
 */
void getopt_url(const char *optarg, const char **url);

/**
 * Parse the option for warning and critical value.
 * \param[in] c commantline switch
 * \param[in] optarg option argument
 * \param[out] threshold threshold to apply range to
 */
void getopt_wc(int c, const char *optarg, thresholds **threshold);

/**
 * Parse the option for warning and critical time range value.
 * \param[in] c commantline switch
 * \param[in] optarg option argument
 * \param[out] threshold threshold to apply range to
 */
void getopt_wc_time(int c, const char *optarg, thresholds **threshold);

/**
 * Parse the option for warning and critical value. This function
 * sets alert_on if no colon is found.
 * \param[in] c commantline switch
 * \param[in] optarg option argument
 * \param[out] threshold threshold to apply range to
 */
void getopt_wc_at(int c, const char *optarg, thresholds **threshold);

/**
 * Parse the option for warning and critical time range value. This
 * function sets alert_on if no colon is found.
 * \param[in] c commantline switch
 * \param[in] optarg option argument
 * \param[out] threshold threshold to apply range to
 */
void getopt_wc_time_at(int c, const char *optarg, thresholds **threshold);

/**
 * Parse the option for IPv4/IPv6 flags.
 * \param[in] c commantline switch
 * \param[out] family IP protocol families to use.
 */
void getopt_46(int c, int *family);

/** optstring for help, version, verbose */
#define MP_OPTSTR_DEFAULT   "hVvt:"
/** longopts option for help, version, verbose */
#define MP_LONGOPTS_DEFAULT {"help", no_argument, NULL, (int)'h'}, \
                            {"version", no_argument, NULL, (int)'V'}, \
                            {"verbose", no_argument, NULL, (int)'v'}, \
                            {"eopt", optional_argument, NULL, (int)MP_LONGOPT_EOPT}, \
                            {"perfdata", no_argument, (int *)&mp_showperfdata, 1}, \
                            {"timeout", required_argument, NULL, (int)'t'}

/** optstring for default notification options */
#define MP_OPTSTR_NOTIFY   MP_OPTSTR_DEFAULT"F:m:"
/** longopts option for default notification options */
#define MP_LONGOPTS_NOTIFY {"help", no_argument, NULL, (int)'h'}, \
                           {"version", no_argument, NULL, (int)'V'}, \
                           {"verbose", no_argument, NULL, (int)'v'}, \
                           {"eopt", optional_argument, NULL, (int)MP_LONGOPT_EOPT}, \
                           {"timeout", required_argument, NULL, (int)'t'}, \
                           {"file", required_argument, NULL, (int)'F'}, \
                           {"message", required_argument, NULL, (int)'m'}

/** optstring for hostname */
#define MP_OPTSTR_HOST      "H:"
/** longopts option for hostname */
#define MP_LONGOPTS_HOST    {"hostname", required_argument, NULL, (int)'H'}

/** optstring for port */
#define MP_OPTSTR_PORT      "P:"
/** longopts option for port */
#define MP_LONGOPTS_PORT    {"port", required_argument, NULL, (int)'P'}

/** optstring for ipv4 and ipv6 */
#ifdef USE_IPV6
#define MP_OPTSTR_IPV       "46"
#else
#define MP_OPTSTR_IPV
#endif
/** longopts option for ipv4 and ipv6 */
#ifdef USE_IPV6
#define MP_LONGOPTS_IPV     {"ipv4", no_argument, NULL, (int)'4'}, \
                            {"ipv6", no_argument, NULL, (int)'6'}
#else
#define MP_LONGOPTS_IPV
#endif

/** optstring for warn and crit */
#define MP_OPTSTR_WC        "w:c:"
/** longopts option for warn and crit */
#define MP_LONGOPTS_WC      {"warning", required_argument, NULL, (int)'w'}, \
                            {"critical", required_argument, NULL, (int)'c'}

/** longopts final */
#define MP_LONGOPTS_END     {0, 0, NULL, 0}

#endif /* _MP_ARGS_H_ */

/* vim: set ts=4 sw=4 et syn=c : */
